### PR TITLE
[Compose] 2120 - image preview activity improvements

### PR DESCRIPTION
**Pull request by filbabic**
WIP More work done on #2120. Specifically with Image options that are shown in the header.

### 🎯 Goal

We're expanding and improving the set of features we support with the ImagePreviewActivity. We want to give the same experience as we do in the core Android SDK.

### 🛠 Implementation details

- Added a header toggle for images in the preview
- Added options to Download/Delete/Show in chat/reply to an image/message
- Added a way to open the image the user tapped on in the MessageList in the Preview

### 🎨 UI Changes

| Before | After |
| --- | --- |
| before | after |

The menus are also visible below.

| Before | After |
| --- | --- |
| options | gallery |

### 🧪 Testing

Open any image (or images) in MessageList. You should see the ellipsis (3 dots) in the header that toggle image options.

### ☑️ Checklist

- [x] PR targets the `develop` branch
- [x] Changelog is updated with client-facing changes
~~- [] New code is covered by unit tests~~
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (docusaurus, tutorial, CMS (task created))
- [x] Reviewers added

### 🎉 GIF

gallery

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/state/imagepreview/ImagePreviewAction.kt
Comment (zsmb13) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/state/imagepreview/ImagePreviewAction.kt:8]: Do these need to be public?

File: stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/MessagesActivity.kt
Comment (zsmb13) [inline stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/MessagesActivity.kt:129]:
```suggestion
                            null -> Unit
```

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (zsmb13) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:157]:
```suggestion
        val startingPosition = if (initialAttachmentPosition !in message.attachments.indices) 0 else initialAttachmentPosition
```

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (zsmb13) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:200]: Great animations here!

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (zsmb13) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:329]: Not worth using a `LazyColumn` for? Wouldn't we get a nicer solution for the spacers between items if we used that?

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (zsmb13) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:450]: An early return could simplify this method `if (attachments.isEmpty()) return`

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (zsmb13) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:463]: This scaling works for an initial implementation, but we should revisit this and add nicer scale/pan control later on.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (zsmb13) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:575]: This icon looks quite weird to me, we should replace it with the one in our design

**Review by zsmb13 — APPROVED** [review]
Left lots of comments, but almost all of them are minor. I'll leave it up to you which ones you want to address before merging and which ones you'll revisit later, we definitely don't have to take care of all these things before shipping these new things.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageItem.kt]: I'll change it to a `tween()` spec and increase the duration. You can LMK if it works once I update!

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageList.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageList.kt]: Good call - I tried both of the approaches and but figured having an item nullcheck woudl be more clear, however this does effectively prevent a second pass of items.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/messages/MessageListViewModel.kt:612]: Sure - I was thinking about it while working on the snippet, but given that messages are not mutable, another option would be to remove an item from the list and add a new one that has the focused state as true.

I can try doing that and if it makes sense, we can go with that approach.

Edit: Given that the list is not mutable, we'd have to transform it to an ArrayList/MutableList and then perform the operations of removing and adding items, so I think this is not worth much at this point.

We could make the "isFocused" flag as a `var` and just change the target item, re-emit the list. Not sure if it's an improvement honestly. 🤷

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:200]: Thanks! It's just basic stuff that's default but flipped over for our use case!

However, I'm starting to add more of these to make the basic interactions a bit nicer ^^

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:450]: I think actually both an early return and a `finish()` call would be nicer, because the user can essentially remove all the images.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageList.kt
Comment (zsmb13) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/messages/list/MessageList.kt]: Efficiency 💯 😄

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:329]: I think that it would be a bit of an overkill here, as we're just looking for a (generally) static list of a few items.

In terms of the spacers - these are dividers, so we'd have to do something similar here anyway.

So I figured it might've been confusing to have a LazyColumn and have all the items loaded eagerly.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:450]: Will add, also added a fix/improvement when trying to delete the last image in the message.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt:463]: In this case it's kind of the default - zoom in 2x, then zoom in to 3x, then zoom out on taps. 

Or you mean the animation?

Comment (zsmb13) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt]: I mean we're just scaling into the middle of the image regardless of where the taps are happening, and then we can't pan around on the zoomed image

Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/imagepreview/ImagePreviewActivity.kt]: Oh okay! That makes sense - I tried to keep it as close to the UI comp implementation which is fairly basic.

In Compose, we can easily add rotations, panning etc if needed.

**Pull request merged**
